File: 2025/day08.cpp
// https://adventofcode.com/2025/day/8

#include "../common/common.h"
#include "../common/rust.h"

#include <algorithm>
#include <limits>
#include <print>
#include <ranges>
#include <set>

using Boxes = Vec<std::array<i64, 3>>;
constexpr let usize_max = std::numeric_limits<usize>::max();

auto parse(String const& filename) -> Boxes {
  return aoc::views::read_lines(filename) |
         stdv::transform(
             [](str line) { return aoc::split_to_array<3, i64>(line, ','); }) |
         aoc::ranges::to<Boxes>();
}

struct DistanceEntry {
  u64 distance;
  usize from;
  usize to;
};

template <usize NUM_CONNECTIONS>
fn solve_case(Boxes const& boxes) -> usize {
  auto distances =
      stdv::cartesian_product(boxes | stdv::enumerate,
                              boxes | stdv::enumerate) |
      stdv::filter([](auto&& ab) {
        let & [ a, b ] = ab;
        return std::get<0>(a) < std::get<0>(b);
      }) |
      stdv::transform([](auto&& ab) {
        let & [ a, b ] = ab;
        let & [ index_p, p ] = a;
        let & [ index_q, q ] = b;
        return DistanceEntry{
            .distance = (static_cast<u64>(aoc::pown(p[0] - q[0], 2)) +
                         static_cast<u64>(aoc::pown(p[1] - q[1], 2)) +
                         static_cast<u64>(aoc::pown(p[2] - q[2], 2))),
            .from = static_cast<usize>(index_p),
            .to = static_cast<usize>(index_q),
        };
      }) |
      aoc::collect_vec<DistanceEntry>();
  stdr::sort(distances, {}, &DistanceEntry::distance);
  // In the beginning each box is its own circuit
  auto circuit_map = Range{0uz, boxes.size()} | aoc::collect_vec<usize>();
  for (let& de :
       distances | stdv::take(std::min(NUM_CONNECTIONS, distances.size()))) {
    let from = circuit_map[de.from];
    let to = circuit_map[de.to];
    if (from == to) {
      // Already in same circuit, do nothing
      continue;
    }
    // Join circuits by renaming old circuit names to new ones
    let old_name = std::max(from, to);
    let new_name = std::min(from, to);
    auto new_count = 0;
    for (auto& current : circuit_map) {
      if (current == old_name) {
        current = new_name;
      }
      if constexpr (NUM_CONNECTIONS == usize_max) {
        new_count += static_cast<usize>(current == new_name);
      }
    }
    if constexpr (NUM_CONNECTIONS == usize_max) {
      if (new_count == boxes.size()) {
        return static_cast<usize>(boxes[de.from][0] * boxes[de.to][0]);
      }
    }
  }
  AOC_ASSERT(NUM_CONNECTIONS < usize_max,
             "This code should be unreachable in part 2");
  // Sum up circuit sizes
  auto circuit_sizes = Vec(boxes.size(), 0);
  for (let circuit : circuit_map) {
    circuit_sizes[circuit] += 1;
  }
  stdr::sort(circuit_sizes, stdr::greater{});
  return circuit_sizes[0] * circuit_sizes[1] * circuit_sizes[2];
}

int main() {
  std::println("Part 1");
  let example = parse("day08.example");
  AOC_EXPECT_RESULT(40, solve_case<10>(example));
  let input = parse("day08.input");
  AOC_EXPECT_RESULT(102816, solve_case<1000>(input));

  std::println("Part 2");
  AOC_EXPECT_RESULT(25272, solve_case<usize_max>(example));
  AOC_EXPECT_RESULT(100011612, solve_case<usize_max>(input));

  AOC_RETURN_CHECK_RESULT();
}

File: rust/2025/src/day08.rs
use itertools::Itertools;

type Boxes = Vec<(i64, i64, i64)>;

fn parse(filename: &str) -> Boxes {
    let parse = |s: &str| s.parse::<i64>().unwrap();
    aoc::file::read_lines(filename)
        .iter()
        .map(|line| {
            let coords = line.split(',').collect::<Vec<_>>();
            (parse(coords[0]), parse(coords[1]), parse(coords[2]))
        })
        .collect()
}

struct DistanceEntry {
    distance: u64,
    from: usize,
    to: usize,
}

fn solve_case<const NUM_CONNECTIONS: usize>(boxes: &Boxes) -> usize {
    let mut distances = boxes
        .iter()
        .enumerate()
        .cartesian_product(boxes.iter().enumerate())
        .filter(|((index_p, _), (index_q, _))| index_p < index_q)
        .map(
            |(((index_p, &(p1, p2, p3))), (index_q, &(q1, q2, q3)))| DistanceEntry {
                distance: ((p1 - q1).pow(2) as u64
                    + (p2 - q2).pow(2) as u64
                    + (p3 - q3).pow(2) as u64),
                from: index_p,
                to: index_q,
            },
        )
        .collect::<Vec<_>>();
    distances.sort_unstable_by_key(|de| de.distance);
    // In the beginning each box is its own circuit
    let mut circuit_map = (0..boxes.len()).collect::<Vec<_>>();
    for de in &distances[0..(NUM_CONNECTIONS.min(distances.len()))] {
        let from = circuit_map[de.from];
        let to = circuit_map[de.to];
        if (from == to) {
            // Already in same circuit, do nothing
            continue;
        }
        // Join circuits by renaming old circuit names to new ones
        let old_name = from.max(to);
        let new_name = from.min(to);
        let mut new_count = 0;
        for current in circuit_map.iter_mut() {
            if (*current == old_name) {
                *current = new_name;
            }
            new_count += (*current == new_name) as usize;
        }
        if ((NUM_CONNECTIONS == usize::MAX) && (new_count == boxes.len())) {
            return (boxes[de.from].0 * boxes[de.to].0) as usize;
        }
    }
    assert!(
        NUM_CONNECTIONS < usize::MAX,
        "This code should be unreachable in part 2"
    );
    // Sum up circuit sizes
    let mut circuit_sizes = vec![0; boxes.len()];
    for circuit in circuit_map {
        circuit_sizes[circuit] += 1;
    }
    circuit_sizes.sort_unstable();
    circuit_sizes.iter().rev().take(3).product()
}

fn main() {
    println!("Part 1");
    let example = parse("day08.example");
    assert_eq!(40, solve_case::<10>(&example));
    let input = parse("day08.input");
    assert_eq!(102816, solve_case::<1000>(&input));

    println!("Part 2");
    assert_eq!(25272, solve_case::<{ usize::MAX }>(&example));
    assert_eq!(100011612, solve_case::<{ usize::MAX }>(&input));
}
